hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/router/RouterAdminServer.java

```diff
@@ -562,11 +595,35 @@ public GetDestinationResponse getDestination(
       LOG.error("Cannot get location for {}: {}",
           src, ioe.getMessage());
     }
-    if (nsIds.isEmpty() && !locations.isEmpty()) {
-      String nsId = locations.get(0).getNameserviceId();
-      nsIds.add(nsId);
+    return nsIds;
+  }
+
+  /**
+   * Verify the file exists in destination nameservices to avoid dangling
+   * mount points.
+   *
+   * @param entry the new mount points added, could be from add or update.
+   * @return destination nameservices where the file doesn't exist.
+   * @throws IOException
+   */
+  private List<String> verifyFileInDestinations(MountTable entry)
```

Comment:
I would unit tests just for this function if possible.

Reply:
Yeah, I can add that.
This logic change will break a lot of current tests listed by yetus above. What is a good practice here? I have two ideas in mind:
1) make the logic based on a command option when adding/updating mount tables.
2) add mock (if possible) like what I did for TestRouterAdmin.java to fake out the dirs in namenodes.
I like 2 since I think this should be the default check, what's your opinion?

Reply:
I would fix those tests to have those paths existing yes.

Reply:
Thanks for the suggestion. I want to involve more people as well since when I started to fix the tests, I found there are quite a few tests targeting the logic of dangling mount points.
@aajisaka Can you share your thoughts as well?

Reply:
@goiri Uploaded an early version of trying to fix all tests. This is pretty tedious work so before I spend more time on this, let me know your thoughts.
There are mainly two types of tests when dealing with mount table:
1. Use mock RouterRpcServer and so on, this way no downstream namenode calls are made. I put the mock as well, see the change for TestRouterAdmin.java
2. Use real downstream namenode interaction, see TestRouterMountTable.java. I created the paths before calling mount points change.

I kept thinking a much easier way is to add a Router server side config to turn this on and the default is on. In the tests I can just turn the config off explicitly and this way I don't need to deal with individual tests.

Reply:
Eventually we may want to make all the tests be correct, but for now I'm fine setting up the config for just the new tests.

Reply:
If that's the case, I will try to fix all in one batch.

Reply:
Had a discussion with @ayushtkn in the jira ticket and he suggested just making the config to preserve the current workflow and enable this feature optionally. I updated the diff as that and it is not necessary to fix a lot of tests in that case.